runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java

```diff
@@ -526,11 +568,65 @@ private void emitAllPushedBackData() throws Exception {
 
     setPushedBackWatermark(BoundedWindow.TIMESTAMP_MAX_VALUE.getMillis());
 
-    pushbackDoFnRunner.finishBundle();
+  }
+
+  /**
+   * Check whether invoke startBundle, if it is, need to flush the buffer data of
+   * {@link BufferedOutputManager} first.
+   *
+   * <p>In order to avoid having {@link DoFnRunner#processElement(WindowedValue)} or
+   * {@link DoFnRunner#onTimer(String, BoundedWindow, Instant, TimeDomain)} not between
+   * StartBundle and FinishBundle, this method needs to be called in each processElement
+   * and each processWatermark and onProcessingTime. Do not need to call in onEventTime,
+   * because it has been guaranteed in the processWatermark.
+   */
+  private void checkInvokeStartBundle() {
+    if (!bundleStarted) {
+      outputManager.flushBuffer();
```

Comment:
Maybe comment that this call will output elements that were buffered as part of finishing a bundle in snapshot().